chasm/tree.go

```diff
@@ -344,6 +354,183 @@ func (n *Node) setSerializedNode(
 	childNode.setSerializedNode(nodePath[1:], serializedNode)
 }
 
+// serialize sets or updates serializedValue field of the node n with serialized value.
+func (n *Node) serialize() error {
+	switch n.serializedNode.GetMetadata().GetAttributes().(type) {
+	case *persistencespb.ChasmNodeMetadata_ComponentAttributes:
+		return n.serializeComponentNode()
+	case *persistencespb.ChasmNodeMetadata_DataAttributes:
+		return n.serializeDataNode()
+	case *persistencespb.ChasmNodeMetadata_CollectionAttributes:
+		panic("not implemented")
+	case *persistencespb.ChasmNodeMetadata_PointerAttributes:
+		panic("not implemented")
+	default:
+		return serviceerror.NewInternal("unknown node type")
+	}
+}
+
+func (n *Node) serializeComponentNode() error {
+	nodeValueT := reflect.TypeOf(n.value)
+	nodeValueV := reflect.ValueOf(n.value)
+
+	protoMessageFound := false
+	// TODO: consider using walker pattern to unify walking over reflected fields.
+	for i := 0; i < nodeValueT.Elem().NumField(); i++ {
+		fieldV := nodeValueV.Elem().Field(i)
+		fieldT := fieldV.Type()
+		if !fieldT.AssignableTo(protoMessageT) {
+			continue
+		}
+
+		if fieldV.IsNil() {
+			continue
+		}
+
+		if protoMessageFound {
+			return serviceerror.NewInternal("only one proto field allowed in component")
+		}
+		protoMessageFound = true
+
+		blob, err := serialization.ProtoEncodeBlob(fieldV.Interface().(proto.Message), enumspb.ENCODING_TYPE_PROTO3)
+		if err != nil {
+			return err
+		}
+
+		rc, ok := n.registry.componentFor(n.value)
+		if !ok {
+			return serviceerror.NewInternal(fmt.Sprintf("component type %s is not registered", nodeValueT.String()))
+		}
+
+		n.serializedNode.Data = blob
```

Comment:
so we guarantee `n.serializedNode` is not nil here right? even for newly created Fields?
and ^ is guaranteed by `syncSubComponents` below? So when later implementing CloseTransaction, we will call `syncSubComponents` and then `serializeComponentNode`?

Reply:
yes. `serializedNode` guaranteed to be not nil for top level component and for all sub component `syncSubComponents` should be called first. I split "sync" and "serialize" into two functions, because "sync" is recursive and applicable only to top level node, and "serialize" should be called individually on every node.